File: src/main/scala/simann/Annealing.scala
package simann

import scalaz.Scalaz._
import scalaz.{Monad, State}

trait Annealable[A] {
  def heat(a: A): State[Random, A]
  def energy(a: A): Double
}

case class Temperature(value: Double) extends AnyVal {
  def *(x: Double) = Temperature(value * x)
}

object Looping {
  def loopM[F[_], A](start: A)(size: Int, f: A => F[A])(implicit M: Monad[F]): F[A] = size match {
    case 0 => M.pure(start)
    case n => loopM(start)(n - 1, f).flatMap(f)
  }

  def loopS[S, A](start: A)(size: Int, f: A => State[S, A]): State[S, A] = loopM[({type M[B] = State[S, B]})#M, A](start)(size, f)
}

object Annealing {
  def anneal[A](start: A, temperatures: Stream[Temperature], stepsAtEachTemperature: Int)(implicit annealable: Annealable[A]): State[Random, Option[A]] = {
    val trialSolutions: Stream[State[Random, A]] = temperatures.map { temperature => Looping.loopS(start)(stepsAtEachTemperature, testTrialSolution(temperature) _) }
    val sequenced = trialSolutions.sequenceU
    sequenced.map(_.find(a => annealable.energy(a) === 0))
  }

  def temperatures(initialTemperature: Temperature, finalTemperature: Temperature, temperatureDropRatio: Double): Stream[Temperature] =
    unfold(initialTemperature)(t => (t.value > finalTemperature.value) ? (t, (t * temperatureDropRatio)).some | None)

  private def acceptanceProbability(temperature: Temperature) = (d: Double) => math.exp(-d / temperature.value)

  private def testTrialSolution[A: Annealable](temperature: Temperature)(current: A): State[Random, A] = {
    val annealable = implicitly[Annealable[A]]
    val trialSolution = annealable.heat(current)
    for {
      ts <- trialSolution
      threshold <- Random.nextDouble
    } yield {
      choose(ts, current, acceptanceProbability(temperature), threshold, annealable.energy _)
    }
  }

  def choose[A](trialSolution: A, current: A, acceptanceProbability: Double => Double, threshold: => Double, energy: A => Double): A = {
    val improvement = energy(trialSolution) - energy(current)
    if (improvement < 0) {
      trialSolution
    } else {
      val calc = acceptanceProbability(improvement)
      (calc > threshold) ? trialSolution | current
    }
  }
}

File: src/main/hs/Annealing.hs
{-# LANGUAGE ScopedTypeVariables #-}

module Annealing where

import Control.Monad ((>>=), sequence)
import Control.Monad.Random (Rand, Random, RandomGen, getRandomR)
import Data.List (unfoldr, find)

class Annealable a where
    heat :: (RandomGen g) => a -> Rand g a
    energy :: a -> Double

newtype Temperature = Temperature Double deriving (Eq, Show)

loopM :: (Monad m) => a -> Int -> (a -> m a) -> m a
loopM a 0 _ = return a
loopM a n f = (loopM a (n-1) f) >>= f 

anneal :: (RandomGen g, Annealable a) => a -> [Temperature] -> Int -> Rand g (Maybe a)
anneal start temperatures stepsAtEachTemperature = fmap (\x -> find (\a -> energy a == 0) x) trialSolutions
    where trialSolutions = listTrialSolutions start temperatures stepsAtEachTemperature

temperatures :: Temperature -> Temperature -> Double -> [Temperature]
temperatures it@(Temperature initialTemperature) (Temperature finalTemperature) temperatureDropRatio = temperatures
    where step t@(Temperature tv) = if tv > finalTemperature then Just $ (t, Temperature $ tv * temperatureDropRatio) else Nothing
          temperatures = unfoldr step it

listTrialSolutions :: (RandomGen g, Annealable a) => a -> [Temperature] -> Int -> Rand g [a]
listTrialSolutions start temperatures stepsAtEachTemperature = 
    let choose' temperature trialSolution current threshold = choose trialSolution current threshold (acceptanceProbability temperature) energy in 
    sequence $ map (\temperature -> loopM start stepsAtEachTemperature (testTrialSolution $ choose' temperature)) temperatures

testTrialSolution :: (Annealable a, RandomGen g) => (a -> a -> Double -> a) -> a -> Rand g a
testTrialSolution choose current = do
    trialSolution <- heat current
    threshold <- getRandomR ((0.0 :: Double), 1.0)
    return $ choose trialSolution current threshold

acceptanceProbability :: Temperature -> (Double -> Double)
acceptanceProbability (Temperature t) d = exp ((-d) / t)

choose :: a -> a -> Double -> (Double -> Double) -> (a -> Double) -> a
choose trialSolution current threshold acceptanceProbability energy = 
    let improvement = (energy trialSolution) - (energy current) in
    if improvement < 0 then
        trialSolution
    else
        let calc = acceptanceProbability improvement in
        if calc > threshold then trialSolution else current

